File: deno-knowledge-app/src/collectors/global.ts
// collect_global — mirrors knowledge.py collect_global
// Inspects ~/.claude: CLAUDE.md, skills/, settings.json hooks (keys only, no env values), agents/

import { readText } from "../shared.ts";
import { estTokens, frontmatterField, mdHeaders } from "../md.ts";
import { join } from "@std/path";

const SCRIPT_EXTS = new Set([
  ".py",
  ".sh",
  ".bash",
  ".zsh",
  ".js",
  ".mjs",
  ".cjs",
  ".ts",
  ".rb",
  ".pl",
  ".lua",
]);

interface ScriptMeta {
  path: string;
  size: number;
  lang: string;
}

interface SkillMeta {
  name: string;
  description: string;
  tokens: number;
  size_bytes: number;
  has_md: boolean;
  scripts: ScriptMeta[];
}

interface AgentMeta {
  name: string;
  tokens: number;
  size_bytes: number;
  description: string;
}

async function scanScripts(root: string, limit = 40): Promise<ScriptMeta[]> {
  const out: ScriptMeta[] = [];
  try {
    for await (const entry of Deno.readDir(root)) {
      if (out.length >= limit) break;
      const fullPath = join(root, entry.name);
      if (entry.isDirectory) {
        const sub = await scanScripts(fullPath, limit - out.length);
        out.push(...sub);
      } else if (entry.isFile) {
        const dotIdx = entry.name.lastIndexOf(".");
        const ext = dotIdx >= 0 ? entry.name.slice(dotIdx) : "";
        if (!SCRIPT_EXTS.has(ext.toLowerCase())) continue;
        try {
          const stat = await Deno.stat(fullPath);
          const rel = fullPath.slice(root.length + 1);
          out.push({ path: rel, size: stat.size, lang: ext.slice(1) });
        } catch {
          // skip
        }
      }
    }
  } catch {
    // dir not readable
  }
  return out;
}

async function skillMeta(skillDir: string): Promise<SkillMeta> {
  const name = skillDir.split("/").pop()!;
  const mdPath = join(skillDir, "SKILL.md");
  const scripts = await scanScripts(skillDir);
  const base: SkillMeta = {
    name,
    description: "",
    tokens: 0,
    size_bytes: 0,
    has_md: false,
    scripts,
  };
  try {
    const stat = await Deno.stat(mdPath);
    const text = await readText(mdPath);
    if (!text) return base;
    return {
      name,
      description: frontmatterField(text, "description"),
      tokens: estTokens(text),
      size_bytes: stat.size,
      has_md: true,
      scripts,
    };
  } catch {
    return base;
  }
}

async function agentMeta(agentMd: string): Promise<AgentMeta> {
  const filename = agentMd.split("/").pop()!;
  const name = filename.replace(/\.md$/, "");
  try {
    const stat = await Deno.stat(agentMd);
    const text = await readText(agentMd);
    if (!text) return { name, tokens: 0, size_bytes: 0, description: "" };
    return {
      name,
      tokens: estTokens(text),
      size_bytes: stat.size,
      description: frontmatterField(text, "description"),
    };
  } catch {
    return { name, tokens: 0, size_bytes: 0, description: "" };
  }
}

export async function collectGlobal(claudeHome: string): Promise<Record<string, unknown>> {
  try {
    let stat: Deno.FileInfo;
    try {
      stat = await Deno.stat(claudeHome);
    } catch {
      return { available: false, reason: `~/.claude not found: ${claudeHome}` };
    }
    if (!stat.isDirectory) {
      return { available: false, reason: `~/.claude not found: ${claudeHome}` };
    }

    const data: Record<string, unknown> = { available: true, home: claudeHome };

    // CLAUDE.md
    const claudeMdPath = join(claudeHome, "CLAUDE.md");
    const claudeMdText = await readText(claudeMdPath);
    if (claudeMdText !== null) {
      let size = 0;
      try {
        size = (await Deno.stat(claudeMdPath)).size;
      } catch { /* ok */ }
      const managed = /<!--\s*BEGIN cc-setup\s*-->/.test(claudeMdText) &&
        /<!--\s*END cc-setup\s*-->/.test(claudeMdText);
      data.claude_md = {
        size_bytes: size,
        tokens: estTokens(claudeMdText),
        headers: mdHeaders(claudeMdText, [2]),
        managed_block: managed,
      };
    } else {
      data.claude_md = null;
    }

    // skills/
    const skillsDir = join(claudeHome, "skills");
    const skillDirs: string[] = [];
    try {
      for await (const e of Deno.readDir(skillsDir)) {
        if (e.isDirectory && !e.name.startsWith(".") && !e.name.startsWith("_")) {
          skillDirs.push(join(skillsDir, e.name));
        }
      }
      skillDirs.sort();
    } catch { /* no skills dir */ }
    const skillItems = await Promise.all(skillDirs.map(skillMeta));
    const skillNames = skillDirs.map((d) => d.split("/").pop()!);
    data.skills = { count: skillNames.length, names: skillNames, items: skillItems };

    // settings.json hooks — event names, matcher, type, command. env NEVER read.
    const settingsPath = join(claudeHome, "settings.json");
    const settingsText = await readText(settingsPath);
    if (settingsText !== null) {
      try {
        const raw = JSON.parse(settingsText) as Record<string, unknown>;
        const hooks = (raw.hooks ?? {}) as Record<string, unknown>;
        const events: Record<string, number> = {};
        const detail: Record<string, Array<Record<string, string>>> = {};
        if (typeof hooks === "object") {
          for (const [eventName, matchers] of Object.entries(hooks)) {
            let n = 0;
            const entries: Array<Record<string, string>> = [];
            if (Array.isArray(matchers)) {
              for (const matcher of matchers) {
                if (!matcher || typeof matcher !== "object") continue;
                const pattern = String((matcher as Record<string, unknown>).matcher ?? "");
                const inner = (matcher as Record<string, unknown>).hooks;
                if (!Array.isArray(inner)) continue;
                for (const h of inner) {
                  if (!h || typeof h !== "object") continue;
                  n++;
                  entries.push({
                    matcher: pattern,
                    type: String((h as Record<string, unknown>).type ?? ""),
                    command: String((h as Record<string, unknown>).command ?? ""),
                  });
                }
              }
            }
            events[eventName] = n;
            detail[eventName] = entries;
          }
        }
        data.settings = { hook_events: events, hook_detail: detail };
      } catch {
        data.settings = { hook_events: {}, hook_detail: {}, parse_error: true };
      }
    } else {
      data.settings = null;
    }

    // agents/*.md
    const agentsDir = join(claudeHome, "agents");
    const agentFiles: string[] = [];
    try {
      for await (const e of Deno.readDir(agentsDir)) {
        if (e.isFile && e.name.endsWith(".md")) {
          agentFiles.push(join(agentsDir, e.name));
        }
      }
      agentFiles.sort();
    } catch { /* no agents dir */ }
    data.agents = agentFiles.map((f) => f.split("/").pop()!.replace(/\.md$/, ""));
    data.agent_items = await Promise.all(agentFiles.map(agentMeta));

    return data;
  } catch (exc) {
    return { available: false, reason: `collect_global failed: ${exc}` };
  }
}

File: go-knowledge-app/internal/collectors/global.go
package collectors

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"go-knowledge-app/internal/md"
	"go-knowledge-app/internal/shared"
)

var scriptExts = map[string]bool{
	".py": true, ".sh": true, ".bash": true, ".zsh": true, ".js": true,
	".mjs": true, ".cjs": true, ".ts": true, ".rb": true, ".pl": true, ".lua": true,
}

// ScriptMeta describes one script discovered inside a skill directory.
type ScriptMeta struct {
	Path string `json:"path"`
	Size int64  `json:"size"`
	Lang string `json:"lang"`
}

// SkillMeta is one global skill's metadata.
type SkillMeta struct {
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Tokens      int          `json:"tokens"`
	MetaTokens  int          `json:"meta_tokens"`
	SizeBytes   int64        `json:"size_bytes"`
	HasMd       bool         `json:"has_md"`
	Scripts     []ScriptMeta `json:"scripts"`
}

// AgentMeta is one global agent's metadata.
type AgentMeta struct {
	Name        string `json:"name"`
	Tokens      int    `json:"tokens"`
	MetaTokens  int    `json:"meta_tokens"`
	SizeBytes   int64  `json:"size_bytes"`
	Description string `json:"description"`
}

func scanScripts(root string, limit int) []ScriptMeta {
	out := []ScriptMeta{}
	entries, err := os.ReadDir(root)
	if err != nil {
		return out
	}
	for _, entry := range entries {
		if len(out) >= limit {
			break
		}
		fullPath := filepath.Join(root, entry.Name())
		if entry.IsDir() {
			out = append(out, scanScripts(fullPath, limit-len(out))...)
		} else {
			ext := ""
			if dot := strings.LastIndex(entry.Name(), "."); dot >= 0 {
				ext = entry.Name()[dot:]
			}
			if !scriptExts[strings.ToLower(ext)] {
				continue
			}
			info, err := os.Stat(fullPath)
			if err != nil {
				continue
			}
			rel := fullPath[len(root)+1:]
			out = append(out, ScriptMeta{Path: rel, Size: info.Size(), Lang: ext[1:]})
		}
	}
	return out
}

func skillMeta(skillDir string) SkillMeta {
	name := lastPathSegment(skillDir)
	mdPath := filepath.Join(skillDir, "SKILL.md")
	scripts := scanScripts(skillDir, 40)
	base := SkillMeta{Name: name, Scripts: scripts}
	info, err := os.Stat(mdPath)
	if err != nil {
		return base
	}
	text, ok := shared.ReadText(mdPath)
	if !ok || text == "" {
		return base
	}
	description := md.FrontmatterField(text, "description")
	return SkillMeta{
		Name:        name,
		Description: description,
		Tokens:      md.EstTokens(text),
		MetaTokens:  md.EstTokens(name + "\n" + description),
		SizeBytes:   info.Size(),
		HasMd:       true,
		Scripts:     scripts,
	}
}

func agentMeta(agentMd string) AgentMeta {
	filename := lastPathSegment(agentMd)
	name := strings.TrimSuffix(filename, ".md")
	info, err := os.Stat(agentMd)
	if err != nil {
		return AgentMeta{Name: name}
	}
	text, ok := shared.ReadText(agentMd)
	if !ok || text == "" {
		return AgentMeta{Name: name}
	}
	description := md.FrontmatterField(text, "description")
	return AgentMeta{
		Name:        name,
		Tokens:      md.EstTokens(text),
		MetaTokens:  md.EstTokens(name + "\n" + description),
		SizeBytes:   info.Size(),
		Description: description,
	}
}

var beginCcSetupRe = regexp.MustCompile(`<!--\s*BEGIN cc-setup\s*-->`)
var endCcSetupRe = regexp.MustCompile(`<!--\s*END cc-setup\s*-->`)

// CollectGlobal inspects ~/.claude: CLAUDE.md, skills/, settings.json hooks
// (keys only, no env values), agents/.
func CollectGlobal(claudeHome string) map[string]any {
	info, err := os.Stat(claudeHome)
	if err != nil || !info.IsDir() {
		return map[string]any{"available": false, "reason": fmt.Sprintf("~/.claude not found: %s", claudeHome)}
	}

	data := map[string]any{"available": true, "home": claudeHome}

	// CLAUDE.md
	claudeMdPath := filepath.Join(claudeHome, "CLAUDE.md")
	if text, ok := shared.ReadText(claudeMdPath); ok {
		var size int64
		if st, err := os.Stat(claudeMdPath); err == nil {
			size = st.Size()
		}
		managed := beginCcSetupRe.MatchString(text) && endCcSetupRe.MatchString(text)
		data["claude_md"] = map[string]any{
			"size_bytes":    size,
			"tokens":        md.EstTokens(text),
			"headers":       md.MdHeaders(text, []int{2}),
			"managed_block": managed,
		}
	} else {
		data["claude_md"] = nil
	}

	// skills/
	skillsDir := filepath.Join(claudeHome, "skills")
	skillDirs := []string{}
	if entries, err := os.ReadDir(skillsDir); err == nil {
		for _, e := range entries {
			if e.IsDir() && !strings.HasPrefix(e.Name(), ".") && !strings.HasPrefix(e.Name(), "_") {
				skillDirs = append(skillDirs, filepath.Join(skillsDir, e.Name()))
			}
		}
		sort.Strings(skillDirs)
	}
	skillItems := make([]SkillMeta, 0, len(skillDirs))
	skillNames := make([]string, 0, len(skillDirs))
	for _, d := range skillDirs {
		skillItems = append(skillItems, skillMeta(d))
		skillNames = append(skillNames, lastPathSegment(d))
	}
	data["skills"] = map[string]any{"count": len(skillNames), "names": skillNames, "items": skillItems}

	// settings.json hooks — event names, matcher, type, command. env NEVER read.
	if settingsText, ok := shared.ReadText(filepath.Join(claudeHome, "settings.json")); ok {
		var raw map[string]any
		if json.Unmarshal([]byte(settingsText), &raw) != nil {
			data["settings"] = map[string]any{"hook_events": map[string]int{}, "hook_detail": map[string]any{}, "parse_error": true}
		} else {
			events := map[string]int{}
			detail := map[string]any{}
			if hooks, ok := raw["hooks"].(map[string]any); ok {
				for eventName, matchers := range hooks {
					n := 0
					entries := []map[string]string{}
					if arr, ok := matchers.([]any); ok {
						for _, matcher := range arr {
							mm, ok := matcher.(map[string]any)
							if !ok {
								continue
							}
							pattern := asStr(mm["matcher"])
							inner, ok := mm["hooks"].([]any)
							if !ok {
								continue
							}
							for _, h := range inner {
								hm, ok := h.(map[string]any)
								if !ok {
									continue
								}
								n++
								entries = append(entries, map[string]string{
									"matcher": pattern,
									"type":    asStr(hm["type"]),
									"command": asStr(hm["command"]),
								})
							}
						}
					}
					events[eventName] = n
					detail[eventName] = entries
				}
			}
			data["settings"] = map[string]any{"hook_events": events, "hook_detail": detail}
		}
	} else {
		data["settings"] = nil
	}

	// agents/*.md
	agentsDir := filepath.Join(claudeHome, "agents")
	agentFiles := []string{}
	if entries, err := os.ReadDir(agentsDir); err == nil {
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
				agentFiles = append(agentFiles, filepath.Join(agentsDir, e.Name()))
			}
		}
		sort.Strings(agentFiles)
	}
	agentNames := make([]string, 0, len(agentFiles))
	agentItems := make([]AgentMeta, 0, len(agentFiles))
	for _, f := range agentFiles {
		agentNames = append(agentNames, strings.TrimSuffix(lastPathSegment(f), ".md"))
		agentItems = append(agentItems, agentMeta(f))
	}
	data["agents"] = agentNames
	data["agent_items"] = agentItems

	return data
}

// asStr mirrors JS String(x ?? "") for the scalar values read from settings.
func asStr(v any) string {
	if v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}
